File: input/deprecated/lib/quadtree.js
/*
 This Source Code Form is subject to the terms of the Mozilla Public
 License, v. 2.0. If a copy of the MPL was not distributed with this
 file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

class point {
  constructor(shape, userdata) {
    this.shape = shape;
    this.userdata = userdata;
  }
}

class quadtree {
  constructor(boundary, capacity) {
    if (!boundary) {
      throw TypeError('boundary is null or undefined');
    }
    if (!(boundary instanceof rectangle)) {
      throw TypeError('boundary should be a rectangle');
    }
    if (typeof capacity !== 'number') {
      throw TypeError(`capacity should be a number but is a ${typeof capacity}`);
    }
    if (capacity < 1) {
      throw RangeError('capacity must be greater than 0');
    }
    this.boundary = boundary;
    this.capacity = capacity;
    this.points = [];
    this.divided = false;
  }
  subdivide() {
    let x = this.boundary.x;
    let y = this.boundary.y;
    let w = this.boundary.width / 2;
    let h = this.boundary.height / 2;

    let nw = new rectangle(new pos(x + w, y, 0), w, h, new gradient());
    let ne = new rectangle(new pos(x + w, y + h, 0), w, h, new gradient());
    let se = new rectangle(new pos(x, y + h, 0), w, h, new gradient());
    let sw = new rectangle(new pos(x, y, 0), w, h, new gradient());

    this.northwest = new quadtree(nw, this.capacity);
    this.northeast = new quadtree(ne, this.capacity);
    this.southeast = new quadtree(se, this.capacity);
    this.southwest = new quadtree(sw, this.capacity);

    this.divided = true;
  }

  insert(point) {
    if (!this.boundary.contains(point.shape)) {
      return false;
    }
    if (this.points.length < this.capacity) {
      this.points.push(point);
      return true;
    }
    if (!this.divided) {
      this.subdivide();
    }
    return (
        this.northeast.insert(point) || this.northwest.insert(point) || this.southeast.insert(point) ||
        this.southwest.insert(point));
  }

  query(range, found) {
    if (!found) {
      found = [];
    }
    if (!range.intersects(this.boundary)) {
      return found;
    }
    for (let p of this.points) {
      if (range.contains(p.shape)) {
        found.push(p);
      }
    }
    if (this.divided) {
      this.northwest.query(range, found);
      this.northeast.query(range, found);
      this.southwest.query(range, found);
      this.southeast.query(range, found);
    }
    return found;
  }
}

File: src/util/quadtree.cpp
/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

#include "util/quadtree.h"

point_type::point_type(position pos, int userdata) : pos(pos), userdata(userdata) {}

quadtree::quadtree() : boundary(rectangle_v2(position(0, 0), 0, 0)) {}

quadtree::quadtree(rectangle_v2 boundary, size_t capacity)
    : boundary(boundary), capacity(std::max(capacity, size_t(1))), divided(false) {}

void quadtree::subdivide() {
  auto x = boundary.top_left.x;
  auto y = boundary.top_left.y;
  auto w = boundary.width / 2;
  auto h = boundary.height / 2;

  rectangle_v2 nw(position(x + w, y), w, h);
  rectangle_v2 ne(position(x + w, y + h), w, h);
  rectangle_v2 se(position(x, y + h), w, h);
  rectangle_v2 sw(position(x, y), w, h);

  northwest = std::make_unique<quadtree>(nw, capacity);
  northeast = std::make_unique<quadtree>(ne, capacity);
  southeast = std::make_unique<quadtree>(se, capacity);
  southwest = std::make_unique<quadtree>(sw, capacity);

  divided = true;
}

bool quadtree::insert(point_type point) {
  if (!boundary.contains(point.pos)) {
    return false;
  }
  if (points.size() < capacity) {
    points.push_back(point);
    return true;
  }
  if (!divided) {
    subdivide();
  }
  return (northeast->insert(point) || northwest->insert(point) || southeast->insert(point) || southwest->insert(point));
}

void quadtree::query(size_t index, const circle_v2& range, std::vector<point_type>& found) {
  if (!range.intersects(boundary)) {
    return;
  }
  for (const auto& p : points) {
    if (p.userdata == index) {
      // skip collision with self
      continue;
    }
    if (range.contains(p.pos)) {
      found.push_back(p);
    }
  }
  if (divided) {
    northwest->query(index, range, found);
    northeast->query(index, range, found);
    southwest->query(index, range, found);
    southeast->query(index, range, found);
  }
}
